### PR TITLE
Adds read/write support for v1.1 system symbols

### PR DESCRIPTION
**Pull request by zslayton**
* Adds static v1.1 system symbol table definition
* Adds `SystemSymbol_1_1` variant of `RawSymbolRef`
* Renames old `RawSymbolRef::as_raw_symbol_token_ref` to shorter `as_raw_symbol_ref` for consistency.
* Adds read/write support for `0xEE` opcode, system symbol values
* Adds read/write support for system symbols appearing in `FlexSym`
* The application-level writer now uses system symbols for its encoding directives

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by zslayton — COMMENTED** [review]
🗺️ PR tour 🧭

File: src/lazy/encoder/binary/v1_0/container_writers.rs [hunk 1]
```diff
@@ -52,16 +52,20 @@ impl<'value, 'top> BinaryContainerWriter_1_0<'value, 'top> {
         let iterator = annotations.into_iter();
         let mut symbol_ids = BumpVec::with_capacity_in(iterator.size_hint().0, self.allocator);
         for annotation in iterator {
-            match annotation.as_raw_symbol_token_ref() {
-                RawSymbolRef::SymbolId(symbol_id) => symbol_ids.push(symbol_id),
+            let symbol_address = match annotation.as_raw_symbol_ref() {
+                RawSymbolRef::SymbolId(symbol_id) => symbol_id,
+                RawSymbolRef::SystemSymbol_1_1(_system_symbol) => {
+                    unreachable!("Ion 1.1 symbol in Ion 1.0 writer")
+                }
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/encoder/binary/v1_0/container_writers.rs:59]: I think the best we could do without pervasive generics is have a separate enum with only `SymbolId`/`Text` and 'downgrade' to that whenever we enter a 1.0-only code path.

File: src/lazy/encoder/binary/v1_0/value_writer.rs [hunk 2]
```diff
@@ -220,11 +220,14 @@ impl<'value, 'top> BinaryValueWriter_1_0<'value, 'top> {
...
 
     pub fn write_symbol<A: AsRawSymbolRef>(self, value: A) -> IonResult<()> {
-        match value.as_raw_symbol_token_ref() {
+        match value.as_raw_symbol_ref() {
             RawSymbolRef::SymbolId(sid) => self.write_symbol_id(sid),
             RawSymbolRef::Text(text) => IonResult::illegal_operation(format!(
                 "the Ion 1.0 raw binary writer cannot write text symbols (here: '{text}')"
             )),
+            RawSymbolRef::SystemSymbol_1_1(_) => {
+                unreachable!("Ion 1.1 token in 1.0 value writer")
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/encoder/binary/v1_0/value_writer.rs:229]: It's only `unreachable!` in the sense that this API is private/experimental and this branch would be a bug. I can downgrade this to an illegal operation.

File: src/raw_symbol_ref.rs [hunk 3]
```diff
@@ -17,10 +51,25 @@ impl<'a> RawSymbolRef<'a> {
         match self {
             RawSymbolRef::SymbolId(sid) => symbol_id == *sid,
             RawSymbolRef::Text(text) => symbol_text == *text,
+            // XXX: This method doesn't make much sense now that Ion 1.1 system symbols have their
+            //      own address space.
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/raw_symbol_ref.rs:55]: Yeah, it's just a leaky abstraction. This method superficially ignores the `SystemSymbol_1_1` case, forcing the implementation to make a decision. We can treat it as text, but if there was a system symbol then it was a 1.1 token and the caller may have cared about the distinction between `Text` and `SystemSymbol_1_1`.

File: src/raw_symbol_ref.rs [hunk 4]
```diff
@@ -29,12 +78,14 @@ impl<'a> RawSymbolRef<'a> {
                     #[inline(never)]
                     || {
                         IonError::decoding_error(format!(
-                            "found a symbol ID (${}) that was not in the symbol table",
-                            sid
+                            "found {label} symbol ID (${}) that was not in the symbol table\n{:?}",
+                            sid,
+                            context.symbol_table()
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/raw_symbol_ref.rs:83]: Eep, this was a debugging holdover! Thanks for pointing it out, I'll revert this.

File: src/raw_symbol_ref.rs [hunk 5]
```diff
@@ -123,6 +174,119 @@ impl<'a> From<SymbolRef<'a>> for RawSymbolRef<'a> {
 
 impl<'a> From<&'a Symbol> for RawSymbolRef<'a> {
     fn from(value: &'a Symbol) -> Self {
-        value.as_raw_symbol_token_ref()
+        value.as_raw_symbol_ref()
+    }
+}
+
+pub(crate) trait SystemSymbol: Sized {
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/raw_symbol_ref.rs:181]: Initially I had a `SystemSymbol` struct that held an `(IonVersion, SymbolAddress)` tuple. `SystemSymbolTable` instances would return `SystemSymbol`s, and `RawSymbolRef` had a `SystemSymbol` variant that wasn't Ion-1.1-specific.

However, this led to a lot of situations in Ion 1.0 code where it was unclear whether an Ion 1.0 system symbol would be represented as a `SymbolId` or a `SystemSymbol`. To eliminate that ambiguity, I split `SystemSymbol` into `SystemSymbol_1_1` and `SystemSymbol_1_0` and narrowed `RawSymbolRef::SystemSymbol` to `RawSymbolRef::SystemSymbol_1_1`. Then I turned the `SystemSymbol` type into a trait.

While I think this was a reasonable way to preserve the abstraction, we aren't using the abstraction much in practice right now. In the spirit of YAGNI, I can strip it out for the moment and we can reintroduce it later if/when we have a more concrete use case.

File: src/lazy/encoder/binary/v1_1/flex_sym.rs [hunk 6]
```diff
@@ -49,9 +50,13 @@ impl<'top> FlexSym<'top> {
     /// Encodes the empty string or symbol ID zero as a FlexSym. The caller is responsible for
     /// confirming that `symbol` is one of those two cases before calling.
     fn encode_special_case(output: &mut BumpVec<u8>, symbol: RawSymbolRef) {
+        use RawSymbolRef::*;
         let encoding: &[u8] = match symbol {
-            SymbolId(_) => &[FlexSym::ZERO, 0xE1, 0x00],
-            Text(_) => &[FlexSym::ZERO, 0x90],
+            SymbolId(_) => &[FlexSym::ZERO, 0x60],
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/encoder/binary/v1_1/flex_sym.rs:55]: The method's doc comment explains this, but I've added a comment inline for this and the `Text(_)` branch.

File: src/text/text_formatter.rs [hunk 7]
```diff
@@ -278,25 +278,23 @@ impl<'a, W: std::fmt::Write> FmtValueFormatter<'a, W> {
...
-            }
-            RawSymbolRef::Text(text) => {
-                // Write the symbol text using quotes and escaping any characters that require it.
-                write!(self.output, "\'")?;
-                self.format_escaped_text_body(text)?;
-                write!(self.output, "\'")?;
-            }
+        let text = match token.as_raw_symbol_ref() {
+            RawSymbolRef::SymbolId(sid) => return Ok(write!(self.output, "${sid}")?),
+            RawSymbolRef::SystemSymbol_1_1(system_symbol) => system_symbol.text(),
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/text/text_formatter.rs:283]: Hey, good observation! I hadn't realized there was only one symbol to quote--done.

File: src/types/symbol.rs [hunk 8]
```diff
@@ -98,6 +99,12 @@ impl Symbol {
         }
     }
 
+    pub fn system(text: &'static str) -> Symbol {
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/types/symbol.rs:102]: I originally named this method `static` but... that's a keyword :(

I renamed it `static_text`--howzat?

File: src/lazy/encoder/writer.rs [hunk 9]
```diff
@@ -873,11 +884,11 @@ mod tests {
             ],
             &[
                 0xE9, // Opcode: FlexUInt follows with byte length of sequence
-                0x09, // FlexUInt byte length: 4
+                0x0B, // FlexUInt byte length: 4
                 0x07, // FlexSym SID $3
                 0x09, // FlexSym SID $4
                 0x0D, // FlexSym SID $6
-                0x15, // FlexSym SID $10
+                0x0A, 0x01, // FlexSym SID $66
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/encoder/writer.rs:891]: If these were tests we expected to update periodically, it'd probably be worth the effort. Hopefully these are stable now though, and more extensive future verification will be done via the DSL in `ion-tests`.

File: src/raw_symbol_ref.rs [hunk 10]
```diff
@@ -123,6 +174,119 @@ impl<'a> From<SymbolRef<'a>> for RawSymbolRef<'a> {
...
+        Self(symbol_address)
+    }
+
+    pub const fn address(&self) -> SymbolAddress {
+        self.0
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) struct SystemSymbol_1_0(SymbolAddress);
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/raw_symbol_ref.rs]: It was here for completeness, but I've removed it now.

File: src/raw_symbol_ref.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (popematt) [reply src/raw_symbol_ref.rs]: I'm not opposed to having it, necessarily, but if we're adding something that isn't used yet, we should probably document that in the code.

File: src/text/text_formatter.rs [hunk 11]
```diff
@@ -278,24 +279,29 @@ impl<'a, W: std::fmt::Write> FmtValueFormatter<'a, W> {
...
-            RawSymbolRef::SymbolId(sid) => return Ok(write!(self.output, "${sid}")?),
-            RawSymbolRef::SystemSymbol_1_1(system_symbol) => system_symbol.text(),
-            RawSymbolRef::Text(text) => text,
+        use RawSymbolRef::*;
+        let write_result = match token.as_raw_symbol_ref() {
+            SymbolId(sid) => write!(self.output, "${sid}"),
+            // '' is the only system symbol that requires quoting; the rest are identifiers.
+            SystemSymbol_1_1(v1_1::system_symbols::EMPTY_TEXT) => write!(self.output, "\'\'"),
+            // Any other system symbol is an identifier and doesn't require quoting.
+            SystemSymbol_1_1(symbol) => return Ok(write!(self.output, "{}", symbol.text())?),
```
Comment (popematt) [inline src/text/text_formatter.rs:288]: Now that it's been refactored to have the `match` expression return a write result, you could get rid of the early return here, right?
```suggestion
            SystemSymbol_1_1(symbol) => write!(self.output, "{}", symbol.text()),
```

**Review by popematt — APPROVED** [review]

**Review by jobarr-amzn — APPROVED** [review]

**Pull request merged**